### PR TITLE
ENH: add NA scalar for missing value indicator, use in StringArray.

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
xref https://github.com/pandas-dev/pandas/issues/28095, https://github.com/pandas-dev/pandas/issues/28778

This PR adds a `pd.NA` singleton with the behaviour as discussed in above issues. For now, it's only used in StringArray in this PR.

### CODE REVIEW

File: pandas/core/na_scalar.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+import numbers
+
+
+def _create_binary_propagating_op(name):
+    def method(self, other):
+        if isinstance(other, numbers.Number) or other is NA or isinstance(other, str):
```
Comment (jorisvandenbossche) [inline pandas/core/na_scalar.py]: Question is what type of objects we should recognize here. 

I think we want to recognize scalar values that can be stored in dataframes. But of course we don't have control over what kind of scalars people put in a dataframe. 
For now I did numbers + string (numbers also covers numpy scalars). Our own scalars (Timedelta, Timestamp, etc) could also be added, or they can also handle that on their side)

File: pandas/core/na_scalar.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    def __str__(self) -> str:
+        return "NA"
+
+    def __bool__(self):
+        raise TypeError("boolean value of NA is ambiguous")
+
+    def __hash__(self):
+        # TODO what should we use here to hash?
+        return 0
```
Comment (jorisvandenbossche) [inline pandas/core/na_scalar.py]: No idea how this should be best implemented. Take a hash of the `self.__dict__` ?

File: pandas/tests/scalar/test_na_scalar.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        assert (NA != other) is NA
+        assert (NA > other) is NA
+        assert (NA >= other) is NA
+        assert (NA < other) is NA
+        assert (NA <= other) is NA
+
+        if isinstance(other, np.int64):
+            # for numpy scalars we get a deprecation warning and False as result
+            # for equality or error for larger/lesser than
+            continue
```
Comment (jorisvandenbossche) [inline pandas/tests/scalar/test_na_scalar.py]: So numpy scalars we don't have full control over, so this means that if they are the left operand, we get some other behaviour:

```
In [27]: np.int64(1) == pd.NA 
/home/joris/miniconda3/envs/dev/bin/ipython:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  #!/home/joris/miniconda3/envs/dev/bin/python
Out[27]: False

In [28]: pd.NA == np.int64(1) 
Out[28]: NA

In [29]: np.int64(1) < pd.NA 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-29-87134fac2734> in <module>
----> 1 np.int64(1) < pd.NA

~/scipy/pandas/pandas/core/na_scalar.py in __bool__(self)
     37 
     38     def __bool__(self):
---> 39         raise TypeError("boolean value of NA is ambiguous")
     40 
     41     def __hash__(self):

TypeError: boolean value of NA is ambiguous

In [30]: pd.NA > np.int64(1) 
Out[30]: NA
```

(for the first case, not sure what the behaviour will be once the change in numpy is done)

File: pandas/core/na_scalar.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        return "NA"
+
+    def __str__(self) -> str:
+        return "NA"
+
+    def __bool__(self):
+        raise TypeError("boolean value of NA is ambiguous")
+
+    def __hash__(self):
+        # TODO what should we use here to hash?
```
Comment (TomAugspurger) [inline pandas/core/na_scalar.py]: This is a singleton right? You could use `id(self)` then?

Comment (jorisvandenbossche) [reply pandas/core/na_scalar.py]: Ah, yes. That is also what is used for a default hash (in case you don't implement equality, and thus don't have to implement hash yourself)

File: pandas/tests/scalar/test_na_scalar.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        not NA
+
+
+def test_hashable():
+    assert hash(NA) == hash(NA)
+    d = {NA: "test"}
+    assert d[NA] == "test"
+
+
+def test_arithmetic_ops(all_arithmetic_functions):
```
Comment (Dr-Irv) [inline pandas/tests/scalar/test_na_scalar.py]: I think you should add tests with `np.nan` for both logical and arithmetic ops

Comment (jorisvandenbossche) [reply pandas/tests/scalar/test_na_scalar.py]: Good idea! added

File: pandas/tests/scalar/test_na_scalar.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    op = all_arithmetic_functions
+
+    for other in [NA, 1, 1.0, "a", np.int64(1), np.nan]:
+        if op.__name__ == "rmod" and isinstance(other, str):
+            continue
+        assert op(NA, other) is NA
```
Comment (Dr-Irv) [inline pandas/tests/scalar/test_na_scalar.py]: I'm not 100% sure of this, but do you also need to test the reverse ops, i.e., add
```
                      assert op(other, NA) is NA
```

Comment (jorisvandenbossche) [reply pandas/tests/scalar/test_na_scalar.py]: I think they are included in `all_arithmetic_functions`

Comment (Dr-Irv) [reply pandas/tests/scalar/test_na_scalar.py]: That's why I said I'm not 100% sure.  I know that when I worked on the ops for `ExtensionArray`, I discovered that the way I was leveraging tests for the operators wasn't properly testing the reverse operators, but that was over a year ago, and it involves a different set of tests, so maybe we're OK.

**Comment by TomAugspurger** [comment]
Any chance you could change StringDtype.na_value to use this, and see what breaks?

There are a few places where we use `np.nan` when we should be referring to `StringDtype.na_value` instead.

I suspect that `_libs.lib.Validator.is_valid_null` or `util.is_nan` will need to be updated to recognice pd.NA as NA. So not a big deal if it's a bunch of work to prototype.

File: pandas/core/na_scalar.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/na_scalar.py]: you might want to look at how NaT handles these, in particular for arraylike others

File: pandas/core/na_scalar.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        return "NA"
+
+    def __str__(self) -> str:
+        return "NA"
+
+    def __bool__(self):
+        raise TypeError("boolean value of NA is ambiguous")
+
+    def __hash__(self):
+        return id(self)
```
Comment (jbrockmendel) [inline pandas/core/na_scalar.py]: this is going to be non-deterministic?

File: pandas/core/na_scalar.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    __rand__ = __and__
+
+    def __or__(self, other):
+        if other is True:
+            return True
+        elif other is False or other is NA:
+            return NA
+        else:
+            return NotImplemented
```
Comment (jbrockmendel) [inline pandas/core/na_scalar.py]: other NA objects?  NaT? NaN? None?

**Comment by jbrockmendel** [comment]
core.missing and _libs.missing will need to recognize this

File: pandas/core/na_scalar.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/na_scalar.py]: Do we want it to be recognized by arrays? (I somewhat purposefully left those out, but didn't think it fully through)

For our own internal arrays (eg IntegerArray), that can be handled on the array level? (so returning NotImplemented here)

For numpy arrays the question is what to do with it, as you cannot represent NAs in a numpy array (except in an object array). So not handling it might be fine?

Comment (jorisvandenbossche) [reply pandas/core/na_scalar.py]: > For numpy arrays the question is what to do with it, as you cannot represent NAs in a numpy array (except in an object array). So not handling it might be fine?

Actually, by not handling it and deferring to numpy, this already happens (-> converting to object):

```
In [61]: np.array([1, 2, 3]) + pd.NA  
Out[61]: array([NA, NA, NA], dtype=object)

In [62]: pd.NA + np.array([1, 2, 3]) 
Out[62]: array([NA, NA, NA], dtype=object)
```

File: pandas/core/na_scalar.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/core/na_scalar.py]: > this is going to be non-deterministic?

Since there is only one instance of this class, it will be the same value for each `pd.NA`.

File: pandas/core/na_scalar.py (hunk 8 above)
Comment (jorisvandenbossche) [reply pandas/core/na_scalar.py]: For logical ops, I am not sure it should handle those, as logical ops typically involve booleans (and NaT/NaN/None cannot be stored in a boolean array)

Eg for float is also raises right now:

```
In [69]: pd.NA | 1.5
...
TypeError: unsupported operand type(s) for |: 'NAType' and 'float'

In [70]: pd.NA | np.nan 
...
TypeError: unsupported operand type(s) for |: 'NAType' and 'float'
```

(for comparison ops, though, we should probably add those)

**Comment by TomAugspurger** [comment]
Playing with using this in StringArray, will push something (here or a PR against this) in a bit.

Anecdotally, having `__bool__` raise is kinda annoying since it breaks reductions like `all`. 

```python
(Pdb) pp left_value
<StringArray>
[NA, NA]
Length: 2, dtype: string
(Pdb) np.all(left_value == right_value)
*** TypeError: boolean value of NA is ambiguous
```

But maybe that's a good thing. I'd like to develop some more experience here.

**Comment by jorisvandenbossche** [comment]
@jbrockmendel I moved it to cython (well, still mostly copy paste of the python version into the cython file), so feedback on that part is very welcome

File: pandas/_libs/missing.pyx [hunk 9]
```diff
@@ -44,7 +44,7 @@ cpdef bint checknull(object val):
     The difference between `checknull` and `checknull_old` is that `checknull`
     does *not* consider INF or NEGINF to be NA.
     """
-    return is_null_datetimelike(val, inat_is_null=False)
+    return val is C_NA or is_null_datetimelike(val, inat_is_null=False)
```
Comment (jorisvandenbossche) [inline pandas/_libs/missing.pyx]: I didn't move this yet into is_null_datetimelike (where all other checks happen), as right now the tslibs/nattype do not depend on missing.pyx or something from lib.pyx, and this might be on purpose?

**Comment by jorisvandenbossche** [comment]
> Anecdotally, having __bool__ raise is kinda annoying since it breaks reductions like all.

We will need to rethink any/all in general as well. In principle, those are reducing operations, and reductions skip NAs, so your example should be equivalent to `all` of an empty Series/array. 
Although in the case of any/all, skipping NAs also doesn't fully feel as the best behaviour in many cases (eg for boolean array `all([True, NA]) == True` would be also be somewhat strange as the NA could be False, so you don't know the result). 
This also recently came up for object dtypes IIRC, and was planning to open a new issue for it (or find an existing).

**Comment by TomAugspurger** [comment]
If anyone wants a preview of what getting this working with StringArray looks like: https://github.com/jorisvandenbossche/pandas/pull/1/files.

It'll mean a larger diff, but I think we'll want to merge that PR against Joris' branch into here, so that we can get a feel for what actually using this is going to look like before merging it into master.

File: pandas/core/na_scalar.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/na_scalar.py]: thats fine for now, but longer-term this will need to be optimized

**Comment by jbrockmendel** [comment]
> I suspect that _libs.lib.Validator.is_valid_null or util.is_nan will need to be updated to recognice pd.NA as NA. So not a big deal if it's a bunch of work to prototype.

util.is_nan is specifically for float-nan.  The scalar-recognizing functions that need to recognize this are in _libs.missing.

File: pandas/_libs/missing.pyx (hunk 9 above)
Comment (jbrockmendel) [reply pandas/_libs/missing.pyx]: > right now the tslibs/nattype do not depend on missing.pyx or something from lib.pyx, and this might be on purpose?

Very much on purpose.  missing.pyx depends on nattype.pyx, and im saying it will need to depend on [wherever you're defining pd.NA] too.

Comment (jbrockmendel) [reply pandas/_libs/missing.pyx]: (now that ti look at this in the context of the diff its clear you already did what i had in mind)

File: pandas/_libs/missing.pyx [hunk 10]
```diff
@@ -278,3 +278,126 @@ cdef inline bint is_null_period(v):
...
     # excluding np.datetime64('nat') and np.timedelta64('nat')
     return checknull_with_nat(v)
+
+
+# -----------------------------------------------------------------------------
+# Implementation of NA singleton
+
+
+def _create_binary_propagating_op(name):
+    import numbers
```
Comment (jbrockmendel) [inline pandas/_libs/missing.pyx]: import at top?

File: pandas/core/na_scalar.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/na_scalar.py]: >  but longer-term this will need to be optimized

Can you explain what you mean with this?

File: pandas/_libs/missing.pyx (hunk 9 above)
Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: > Very much on purpose. missing.pyx depends on nattype.pyx, and im saying it will need to depend on [wherever you're defining pd.NA] too.

The long term problem, though, is to have `is_null_datetimelike` to recognize NA, it will need to depend on something (eg missing.pyx) outside of tslibs (as it would be strange to define NA in tslibs). 
Now, in practice, `is_null_datetimelike` is only used in period.pyx, so not a big priority to have NA working there.

File: pandas/_libs/missing.pyx (hunk 10 above)
Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: > import at top?

This is a python import, so was thinking to keep it here for that reason? (the top imports are only cimports)

That said, there might be a cython / cpython API to do the same?

**Comment by TomAugspurger** [comment]
Joris merged in my changes to make StringArray use `pd.NA`. It wasn't too large of a change so far.

I want to call out one behavior change we may want to make: Right now, `Series[string].str` methods that return numeric output (like `.str.count`) return either int or float dtype, depending on whether there are NAs. See https://github.com/pandas-dev/pandas/pull/29597/files#diff-683f8dacd08abda4913680fafe3a4ea7R3512.

I'm proposing that we change that behavior to always return an integer-na dtype in those cases for `string` dtype (no change to object dtype).

**Comment by jorisvandenbossche** [comment]
> I'm proposing that we change that behavior to always return an integer-na dtype in those cases for string dtype (no change to object dtype).

+1 

It's also not a "breaking" change, as string dtype is still new. In general, with those new dtypes I think we should try to return as much as possible also new dtypes for the result (eg also for boolean results, once boolean array has landed)

File: pandas/_libs/missing.pyx (hunk 10 above)
Comment (jreback) [reply pandas/_libs/missing.pyx]: this should be at the top, likely cython just *works* correct with the numbers module (and makes it an optimized call), even w/o cimport.

File: pandas/_libs/missing.pyx [hunk 11]
```diff
@@ -278,3 +278,126 @@ cdef inline bint is_null_period(v):
...
+
+
+cdef class C_NAType:
+    pass
+
+
+class NAType(C_NAType):
+    """
+    NA ("not available") missing value indicator.
+    """
```
Comment (jreback) [inline pandas/_libs/missing.pyx]: can you indicate that this is a singleton

File: pandas/_libs/missing.pyx [hunk 12]
```diff
@@ -278,3 +278,126 @@ cdef inline bint is_null_period(v):
...
+        return NAType._instance
+
+    def __repr__(self) -> str:
+        return "NA"
+
+    def __str__(self) -> str:
+        return "NA"
+
+    def __bool__(self):
+        raise TypeError("boolean value of NA is ambiguous")
```
Comment (jreback) [inline pandas/_libs/missing.pyx]: test for this?

File: pandas/_libs/missing.pyx [hunk 13]
```diff
@@ -278,3 +278,126 @@ cdef inline bint is_null_period(v):
...
+
+    method.__name__ = name
+    return method
+
+
+def _create_unary_propagating_op(name):
+    def method(self):
+        return NA
+
+    method.__name__ = name
```
Comment (jreback) [inline pandas/_libs/missing.pyx]: can you add a doc-string for these? (auto-construct is ok)

File: pandas/_libs/testing.pyx [hunk 14]
```diff
@@ -180,13 +180,15 @@ cpdef assert_almost_equal(a, b,
         # classes can't be the same, to raise error
         assert_class_equal(a, b, obj=obj)
 
-    if a == b:
-        # object comparison
-        return True
     if isna(a) and isna(b):
         # TODO: Should require same-dtype NA?
```
Comment (jbrockmendel) [inline pandas/_libs/testing.pyx]: this might be a reasonable time to start enforcing same-NA

File: pandas/core/arrays/numpy_.py [hunk 15]
```diff
@@ -278,6 +278,9 @@ def fillna(self, value=None, method=None, limit=None):
         return new_values
 
     def take(self, indices, allow_fill=False, fill_value=None):
+        if fill_value is None:
+            # Primarily for subclasses
+            fill_value = self.dtype.na_value
```
Comment (jbrockmendel) [inline pandas/core/arrays/numpy_.py]: why is this necessary?

File: pandas/core/arrays/string_.py [hunk 16]
```diff
@@ -50,16 +47,8 @@ class StringDtype(ExtensionDtype):
...
-    def na_value(self) -> "Scalar":
-        """
-        StringDtype uses :attr:`numpy.nan` as the missing NA value.
-
-        .. warning::
-
-           `na_value` may change in a future release.
-        """
-        return np.nan
+    #: StringDtype.na_value uses pandas.NA
```
Comment (jbrockmendel) [inline pandas/core/arrays/string_.py]: is the colon here meaningful or a typo?

File: pandas/core/arrays/string_.py [hunk 17]
```diff
@@ -205,9 +200,9 @@ def __setitem__(self, key, value):
 
         # validate new items
         if scalar_value:
-            if scalar_value is None:
-                value = np.nan
-            elif not (isinstance(value, str) or np.isnan(value)):
+            if isna(value):
```
Comment (jbrockmendel) [inline pandas/core/arrays/string_.py]: do we want to include _any_ na value?  e.g NaT seems like it should be excluded

File: pandas/tests/test_strings.py [hunk 18]
```diff
@@ -3507,3 +3507,11 @@ def test_string_array(any_string_method):
         assert all(result[columns].dtypes == "string")
         result[columns] = result[columns].astype(object)
     tm.assert_equal(result, expected)
+
+
+@pytest.mark.xfail(reason="not implmented yet")
```
Comment (jbrockmendel) [inline pandas/tests/test_strings.py]: implmented -> implemented

File: pandas/_libs/missing.pyx (hunk 10 above)
Comment (jbrockmendel) [reply pandas/_libs/missing.pyx]: i haven't found a cimport-able version of `numbers`, but that would be really nice.  non-pandas pure-python imports can go at the top.  This gets executed at import-time anyway.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/numpy_.py (hunk 15 above)
Comment (TomAugspurger) [reply pandas/core/arrays/numpy_.py]: For StringArray.take. It wasn't needed before since PandasArray uses `np.nan` as its na_value, which is the default used in take algos.

File: pandas/core/arrays/string_.py (hunk 16 above)
Comment (TomAugspurger) [reply pandas/core/arrays/string_.py]: That's how sphinx extracts docs for class attributes: https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoattribute

File: pandas/core/arrays/string_.py (hunk 17 above)
Comment (TomAugspurger) [reply pandas/core/arrays/string_.py]: It's not clear to me. We at least want to handle None, NA, and nan here. You'd prefer raising on NaT?

Comment (jorisvandenbossche) [reply pandas/core/arrays/string_.py]: I am personally fine with not allowing pd.NaT (it's clearly specific to time-like data). 

But we should then have a helper function for it (or we probably already have one somewhere), or have it as an option in `isna`, as previously (in the PR to my fork), this was:

```
            if (
                value is libmissing.NA
                or value is None
                or (
                    is_float_dtype(value)
                    and not isinstance(value, str)
                    and np.isnan(value)
                )
            ):
```

File: pandas/_libs/missing.pyx (hunk 12 above)
Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: See `test_thruthiness` at https://github.com/pandas-dev/pandas/pull/29597/files#diff-f9e49517f4c860fedd8d5be50921a2acR23

File: pandas/_libs/missing.pyx (hunk 10 above)
Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: OK, moved to the top

File: pandas/_libs/missing.pyx (hunk 11 above)
Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: Expanded the docstring a bit

(still need to add a section to the user guide, and can then link to there as well)

File: pandas/_libs/testing.pyx (hunk 14 above)
Comment (jorisvandenbossche) [reply pandas/_libs/testing.pyx]: Open a separate issue for it?

File: pandas/_libs/missing.pyx (hunk 13 above)
Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: This is only for dunder methods, so not a docstring that users typically see? (eg the Series dunder methods also have no docstring)

**Comment by Dr-Irv** [comment]
@jorisvandenbossche I'm not checking everything but it occurred to me - do the changes you have test things like `fillna()`, `dropna()` and `isna()`, and how do the semantics change with those methods when we used to have `np.nan` mean NA versus `pd.NA`?

**Comment by jorisvandenbossche** [comment]
@Dr-Irv so right now this only implements the NA object itself (and that it is recognized in `isna`), but it's not yet actually used much as missing value indicator. Right now only in the StringArray (which Tom added to this PR). So those things work (and are tested through the string dtype's tests):

```
In [1]: s = pd.Series(['a', pd.NA, 'c'], dtype='string') 

In [2]: s 
Out[2]: 
0      a
1    NaN
2      c
dtype: string

In [3]: s.isna()  
Out[3]: 
0    False
1     True
2    False
dtype: bool

In [4]: s.dropna()
Out[4]: 
0    a
2    c
dtype: string

In [5]: s.fillna('dd') 
Out[5]: 
0     a
1    dd
2     c
dtype: string

In [7]: np.asarray(s)  
Out[7]: array(['a', NA, 'c'], dtype=object)
```

(note the repr still shows it as "NaN", but it's actually pd.NA that is used, that's something we still need to fix)

For those methods, there should be no behavioural changes. It's eg in comparison ops that there is a change (the NA now gets propagated instead of being False):

```
In [8]: s == 'a' 
Out[8]: 
0     True
1      NaN
2    False
dtype: object   # this will become boolean (there is another PR open to add BooleanArray)

In [9]: s.astype(object) == 'a'   # the typical way where strings are stored as object dtype
Out[9]: 
0     True
1    False
2    False
dtype: bool
```

**Comment by jorisvandenbossche** [comment]
(I need to add those examples as docs!)

**Comment by TomAugspurger** [comment]
Merged in master, so this has the new behavior `.str` ops on StringArray returning an IntegerArray.

**Comment by jorisvandenbossche** [comment]
Thanks! I will try to push some docs on pd.NA here.

**Comment by TomAugspurger** [comment]
Do people have thoughts on the behavior of `float(pd.NA)`? We raise right now, which is probably the right move, but we could consider returning `np.nan`. I'm concerned that this conflates NaN and NA even more than we already do.

Right now, if you have a `np.array([0, 1, pd.NA], dtype="object")`, it's a bit hard to get a float dtype ndarray. You would need something like `arr[pd.isna(arr)] = np.nan; arr.astype(float)`

Assuming we don't want `float(NA)` to be `np.nan`, is `np.asarray(integer_array, dtype=float)` enough of a hint that users want to treat NA as NaN?

**Comment by jorisvandenbossche** [comment]
> Do people have thoughts on the behavior of float(pd.NA)

I think raising by default is certainly a reasonable option. Although if we have float array with mask-NA in the future, that might be even more annoying default behaviour there.

(and should add a test for this behaviour)

We should probably add options to `to_numpy` to be more flexible than the single choice of `__array__`? 
For example, could add a `NA_as_NaN` (or `na_as_nan`, `use_nan`, ..) option to indicate you want to use np.nan as the missing value indicator upon conversion to numpy.

**Comment by Dr-Irv** [comment]
> Do people have thoughts on the behavior of `float(pd.NA)`?

As I said on the call, the _semantics_ of `pd.NA` vs. `np.nan` need to be carefully handled in a variety of places.  IMHO, I think we go for the drastic change of saying that `np.nan` will no longer indicate a missing value, and that it only means "not a number" without any option to allow people to do it the "old" way. This also means that `np.isna()` is different than `pd.isna()`.  So in this case, either `float(pd.NA)` returns `pd.NA`, because `pd.NA` has no type, or we raise an exception.

Here's a bit radical idea to make the difference even more obvious, which is instead of using `pd.NA`, we call it `pd.MV` for "missing value", and change the API everywhere such that any keyword or method that uses the "NA" abbreviation is changed to "MV".

**Comment by jorisvandenbossche** [comment]
> So in this case, either `float(pd.NA)` returns pd.NA, because pd.NA has no type, or we raise an exception.

We can in principle control the behaviour of `foat(..)` (with `__float__`), but if we return something, I think it should return a float (that seems like the contract of `float(..)` to me). So then I would raise an error.

> Here's a bit radical idea to make the difference even more obvious, which is instead of using pd.NA, we call it pd.MV for "missing value",

The problem with that is that the big advantage of NA is that it is already used almost everywhere in our API (dropna, fillna, isna, skipna, ...) ..

**Comment by TomAugspurger** [comment]
> We should probably add options to to_numpy to be more flexible than the single choice of __array__?

Yes, I think that's best. Though the question remains: does `asarray(..., dtype="float")` imply `na_as_nan=True`?

If we add an `IntegerArray.to_numpy(dtype, na_as_nan=False)`, what's the interaction between dtype and na_as_nan?

* dtype=float, na_as_nan=T/F: fine
* dtype=int/bool, na_as_nan=T: raise regardless, since NaN isn't an int / bool? Or only if there are NAs?
* dtype=int/bool, na_as_nan=False: raise iff there are missing values, since NA isn't an int / bool?

All-in-all, for now I'm OK with float(NA) raising, and requiring explicit opt-in from users when converting from NA to NaN, via an `*Array.to_numpy(dtype=float, na_as_nan=True)`.

**Comment by Dr-Irv** [comment]
> > Here's a bit radical idea to make the difference even more obvious, which is instead of using pd.NA, we call it pd.MV for "missing value",
> 
> The problem with that is that the big advantage of NA is that it is already used almost everywhere in our API (dropna, fillna, isna, skipna, ...) ..

I get that, but maybe we create new methods (dropmv, fillmv, ismv, skipmv), etc., as the replacements, keep the old ones alive, and just map the "na" named methods to "mv" named methods, and tell people that the "na" based methods will be deprecated in the future.

I realize it's a bit radical, but my sense is that the only reason "na" was used was because numpy used "na", and pandas decided to make `np.nan` mean "missing value", but now's the time to make it really clear that `np.nan` and `pd.MV` are different.

**Comment by jorisvandenbossche** [comment]
> because numpy used "na",

Numpy does not use "na", but "nan" (np.isnan, np.nanmean, etc). I wasn't yet involved at the time, but I suppose the main reason was the terminology from R to use NA.

Personally, I am not convinced that changing all those APIs is worth that.

**Comment by TomAugspurger** [comment]
> but if we return something, I think it should return a float

Python will raise if you don't return a float. So it's either NaN or raise on our own I think.

---

FWIW, I'd like to punt on the interaction of NA and NaN for now. It doesn't directly affect this issue, as we're only using NA in arrays that will not use NaN. And it's a complicated enough discussion that it warrants its own issue.

That said, @Dr-Irv if you think there are things that *are* related to this issue, then continue to raise them. But I think that something like changing the behavior of `isna`, `dropna`, isn't really on the table right now.

**Comment by jorisvandenbossche** [comment]
> FWIW, I'd like to punt on the interaction of NA and NaN for now

Yes, and the conversion to numpy arryay will only come up in a PR to use it in IntegerArray. 
And let's open a separate issue for NA / NaN

**Comment by jorisvandenbossche** [comment]
More comments on the *code* in this PR?

I know I need to fix divmod (should return two NAs), will try to fix that this evening.

**Comment by Dr-Irv** [comment]
> > FWIW, I'd like to punt on the interaction of NA and NaN for now
> 
> Yes, and the conversion to numpy arryay will only come up in a PR to use it in IntegerArray.
> And let's open a separate issue for NA / NaN

Fair enough.

IMHO, I think that converting `pd.NA` to anything (`str`, `float`, `int`) should always raise.  (Probably could debate about `str(pd.NA)`) .  And in terms of arrays, here's some interesting `numpy` behavior:
```
>>> import numpy as np
>>> np.array([5.0, 2.0, None, np.nan], dtype='float')
array([ 5.,  2., nan, nan])
>>> np.array([5.0, 2.0, None, np.nan])
array([5.0, 2.0, None, nan], dtype=object)
>>> np.array([5, 2, None, np.nan], dtype='int')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
>>> np.array([5, 2, np.nan], dtype='int')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: cannot convert float NaN to integer
```

So maybe the way that `numpy` handles `None` is a guide for the behavior we want in pandas when converting to `numpy` arrays?

**Comment by TomAugspurger** [comment]
> More comments on the code in this PR?

In case it helps, see https://github.com/TomAugspurger/pandas/compare/NA-scalar...TomAugspurger:NA-scalar+IntegerArray for integrating this into IntegerArray.

The only change that may be relevant here is the behavior of `int(NA)` and `float(NA)`. Previously, `int(np.nan)` raised a ValueError, since `np.nan` is a float, and some floats can be converted to int, but `np.nan` can't. I think that a TypeError is appropriate since there are no `NA` is the only instance of NAType.

So perhaps just tests for that.

**Comment by TomAugspurger** [comment]
Just to scope this PR, I think that using `pd.NA` in BooleanArray should be done as a followup PR.

File: pandas/_libs/missing.pyx [hunk 19]
```diff
@@ -278,3 +280,137 @@ cdef inline bint is_null_period(v):
...
+        elif other is False or other is NA:
+            return NA
+        else:
+            return NotImplemented
+
+    __ror__ = __or__
+
+    def __xor__(self, other):
+        if other is False or other is True or other is NA:
+            return NA
```
Comment (jbrockmendel) [inline pandas/_libs/missing.pyx]: all of these checks should be for C_NA

File: pandas/_libs/missing.pyx [hunk 20]
```diff
@@ -278,3 +280,137 @@ cdef inline bint is_null_period(v):
...
+
+
+# -----------------------------------------------------------------------------
+# Implementation of NA singleton
+
+
+def _create_binary_propagating_op(name, divmod=False):
+
+    def method(self, other):
+        if isinstance(other, numbers.Number) or other is NA or isinstance(other, str):
```
Comment (jbrockmendel) [inline pandas/_libs/missing.pyx]: can you put the numbers.Number check last, as it will be least performant

File: pandas/_libs/missing.pyx (hunk 19 above)
Comment (TomAugspurger) [reply pandas/_libs/missing.pyx]: Is that a style or performance thing? We alias `NA = C_NA` at https://github.com/pandas-dev/pandas/pull/29597/files#diff-c4b0cf199cc6d5bc2362ceb737159458R416

Comment (jbrockmendel) [reply pandas/_libs/missing.pyx]: performance.  Checking for NA requires a python-space lookup in the module namespace.  Checking for C_NA is a pointer comparison

Comment (TomAugspurger) [reply pandas/_libs/missing.pyx]: Thanks. For my own understanding, that's true even though `NA` is aliased in this module to `C_NA`?

Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: I don't see any notable difference in timings if I go back and forth a few times between this. But it's in the nanoseconds anyway, so probably timing that on my laptop with ``%timeit`` is too noisy to see the effect.

Comment (jbrockmendel) [reply pandas/_libs/missing.pyx]: > Thanks. For my own understanding, that's true even though NA is aliased in this module to C_NA?

Yes.  Nothing prevents a user from doing `pd._libs.missing.NA = 4`, so the compiled cython still has to lookup NA in the namespace every time.  With C_NA the user can't override it, so the check can be optimized.

**Comment by jorisvandenbossche** [comment]
I added some initial docs (whatsnew + section in missing_data

**Comment by Dr-Irv** [comment]
@jorisvandenbossche I've looked at your doc changes.  You may want to say something about that when you read data, the new `pd.NA` type won't be used, until it gets addressed by closing the issue I created #29752 .

File: doc/source/user_guide/missing_data.rst [hunk 21]
```diff
@@ -771,3 +772,132 @@ the ``dtype="Int64"``.
...
+
+.. ipython:: python
+   :okexcept:
+
+   bool(pd.NA)
+
+This also means that ``pd.NA`` cannot be used in a context where it is
+evaluated to a boolean, such as ``if condition: ...`` where ``condition`` can
+potentially be ``pd.NA``. In such cases, :func:`isna` can be used to check
+for ``pd.NA`` or it could be prevented that ``condition`` can be ``pd.NA``
```
Comment (TomAugspurger) [inline doc/source/user_guide/missing_data.rst]:
```suggestion
for ``pd.NA`` or ``condition`` being ``pd.NA`` can be avoided, for example by filling missing values beforehand.
```

(can't apply this directly, since it wraps two lines)

**Review by TomAugspurger — COMMENTED** [review]
Docs look very nice, thanks.

**Comment by jorisvandenbossche** [comment]
cc @pandas-dev/pandas-core more comments on this?

File: doc/source/user_guide/missing_data.rst [hunk 22]
```diff
@@ -771,3 +772,132 @@ the ``dtype="Int64"``.
...
+
+.. warning::
+
+   Experimental: the behaviour of ``pd.NA`` can still change without warning.
+
+.. versionadded:: 1.0.0
+
+Starting from pandas 1.0, an experimental ``pd.NA`` value (singleton) is
+available to represent scalar missing values. At this moment, it is used in
+the nullable integer and boolean data types and the dedicated string data type
```
Comment (jreback) [inline doc/source/user_guide/missing_data.rst]: use commas rather than 3 ands, ideally if you can link to those sections?

File: doc/source/user_guide/missing_data.rst [hunk 23]
```diff
@@ -771,3 +772,132 @@ the ``dtype="Int64"``.
...
+In this case, ``pd.NA`` does not propagate:
+
+.. ipython:: python
+
+   True | False
+   True | pd.NA
+   pd.NA | True
+
+On the other hand, if one of the operands is ``False``, the result depends
+on the value of the other operand. Therefore, in thise case ``pd.NA``
```
Comment (jreback) [inline doc/source/user_guide/missing_data.rst]: thise -> this

File: doc/source/user_guide/missing_data.rst [hunk 24]
```diff
@@ -771,3 +772,132 @@ the ``dtype="Int64"``.
...
+
+   True & True
+   True & False
+   True & pd.NA
+
+
+``NA`` in a boolean context
+---------------------------
+
+Since the actual value of an NA is unknown, it is ambiguous to convert NA
```
Comment (jreback) [inline doc/source/user_guide/missing_data.rst]: I think we have a section gotchas.truth that is very similiar, could link.

**Review by jreback — APPROVED** [review]
lgtm. some very minor doc comments. can you run the asv's to see if isna is impacted in any way here.

**Comment by jorisvandenbossche** [comment]
Updated for the doc comments, thanks! @Dr-Irv I added a sentence on that. But we should probably add a more extensive section on that in a page dedicated to those new data types (for a follow-up).

@jreback I run the `isnull` benchmarks on this PR and on master, nothing seems off out of the error / noise margin.

**Review by TomAugspurger — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jorisvandenbossche nice addition.

unless I missed this, you didn't update nullable integers to return that, assume that's a followon.

**Comment by jorisvandenbossche** [comment]
> you didn't update nullable integers to return that, assume that's a followon.

Yes, as being discussed, see https://github.com/pandas-dev/pandas/issues/29556 for all follow-ups
